Plugins/BuildServerIntegration/VstsAndTfsIntegration/TfsApiHelper.cs

```diff
@@ -37,81 +37,81 @@ private HttpClient InitializeHttpClient(string serverUrl, string projectName, st
+        private async Task<T> HttpGetAsync<T>(string url)
         {
-            var buildDefs = new List<BuildDefinition>();
-
-            using (var response = await _httpClient.GetAsync($"_apis/build/definitions?api-version=2.0"))
+            using (var response = await _httpClient.GetAsync(url))
             {
                 response.EnsureSuccessStatusCode();
                 string json = await response.Content.ReadAsStringAsync();
 
-                var buildDefinitions = JsonConvert.DeserializeObject<ListWrapper<BuildDefinition>>(json);
-
-                if (buildDefinitionNameFilter == null)
-                {
-                    buildDefs.AddRange(buildDefinitions.Value);
-                }
-                else
-                {
-                    foreach (var buildDefinition in buildDefinitions.Value)
-                    {
-                        if (buildDefinitionNameFilter.IsMatch(buildDefinition.Name))
-                        {
-                            buildDefs.Add(buildDefinition);
-                        }
-                    }
-                }
+                return JsonConvert.DeserializeObject<T>(json);
+            }
+        }
+
+        private async Task<string> GetBuildDefinitionsAsync(string buildDefinitionNameFilter)
+        {
+            var isNotFiltered = string.IsNullOrWhiteSpace(buildDefinitionNameFilter);
+            var buildDefinitionUriFilter = isNotFiltered ? string.Empty : "&name=" + buildDefinitionNameFilter;
+            var buildDefinitions = await HttpGetAsync<ListWrapper<BuildDefinition>>(BuildDefinitionsUrl + buildDefinitionUriFilter);
+            if (buildDefinitions.Count != 0)
+            {
+                return GetBuildDefinitionsIds(buildDefinitions.Value);
+            }
+
+            if (isNotFiltered)
+            {
+                return null;
+            }
+
+            buildDefinitions = await HttpGetAsync<ListWrapper<BuildDefinition>>(BuildDefinitionsUrl);
+
+            var tfsBuildDefinitionNameFilter = new Regex(buildDefinitionNameFilter, RegexOptions.Compiled);
+            return GetBuildDefinitionsIds(buildDefinitions.Value.Where(b => tfsBuildDefinitionNameFilter.IsMatch(b.Name)));
+        }
+
+        private string GetBuildDefinitionsIds(IEnumerable<BuildDefinition> buildDefinitions)
+        {
+            if (buildDefinitions != null && buildDefinitions.Any())
+            {
+                return string.Join(",", buildDefinitions.Select(b => b.Id));
             }
 
-            return buildDefs;
+            return null;
         }
 
-        public void ConnectToTfsServer(string serverUrl, string teamCollection, string projectName, string restApiToken, Regex buildDefinitionNameFilter = null)
+        public void ConnectToTfsServer(string serverUrl, string teamCollection, string projectName, string restApiToken, string buildDefinitionNameFilter)
```

Comment:
how often is this method called?
if it is called more than once then we potentially have a problem with `HttpClient` - there would multiple of them and then there is a risk of exhausting available sockets.
Worth a read:
* https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/
* https://blogs.msdn.microsoft.com/alazarev/2017/12/29/disposable-finalizers-and-httpclient/

Ideally we should create only one instance of `HttpClient` in the main codebase and pass it to plugins.

In the immediate case, perhaps it is better to initialise an instance of `HttpClient` in the .ctor like so:

```
public TfsApiHelper(HttpClient client)
{
    _httpClient = httpClient;
}
```

This also makes the class testable (something I would like to see as well).
I typically use https://github.com/richardszalay/mockhttp (related https://stackoverflow.com/a/36427274/2338036).

Reply:
I know that the httpClient should not be disposed but reused, that's why I used only one instance for all the calls but the bad behavior is caused by the Build integration watcher that wrap the plugins.

This method is called at every refresh of the revision grid. Because a new instance of the `VstsAndTfsAdapter` is created each time. That the bad behavior that @afabre and me spoke about in the previous PR.

The previous instance `VstsAndTfsAdapter` is disposed when the new is created and the httpClient is disposed here also (I pushed a fix and debugged to verify that it's really the case...).

>if it is called more than once then we potentially have a problem with HttpClient

>In the immediate case, perhaps it is better to initialize an instance of HttpClient in the .ctor 

>This also makes the class testable 

I could make it for test purpose but that won't change anything for the exhaustion of available sockets because a new instance is created each time that the `VstsAndTfsAdapter` is created.

I think that `BuildServerWatcher` should be change to not create a new instance each time.

Reply:
> I could make it for test purpose but that won't change anything for the exhaustion of available sockets because a new instance is created each time that the VstsAndTfsAdapter is created.

Correct, it won't. But we can test the proposed behaviour.

Reply:
> I think that BuildServerWatcher should be change to not create a new instance each time.

We can deal with it later as it is outside the scope of this PR